cpp/src/community/leiden_impl.cuh

```diff
@@ -713,6 +713,57 @@ std::pair<size_t, weight_t> leiden(
 
   detail::flatten_leiden_dendrogram(handle, graph_view, *dendrogram, clustering);
 
+  // Get unique cluster id
+  size_t local_num_verts = (*dendrogram).get_level_size_nocheck(0);
+  rmm::device_uvector<vertex_t> unique_cluster_ids(local_num_verts, handle.get_stream());
+
+  thrust::copy(handle.get_thrust_policy(),
+               clustering,
+               clustering + local_num_verts,
+               unique_cluster_ids.begin());
+
+  thrust::sort(handle.get_thrust_policy(), unique_cluster_ids.begin(), unique_cluster_ids.end());
+
+  unique_cluster_ids.resize(thrust::distance(unique_cluster_ids.begin(),
+                                             thrust::unique(handle.get_thrust_policy(),
+                                                            unique_cluster_ids.begin(),
+                                                            unique_cluster_ids.end())),
+                            handle.get_stream());
+
+  if constexpr (multi_gpu) {
+    auto recvcounts = cugraph::host_scalar_allgather(
+      handle.get_comms(), unique_cluster_ids.size(), handle.get_stream());
+
+    std::vector<size_t> displacements(recvcounts.size());
+    std::exclusive_scan(recvcounts.begin(), recvcounts.end(), displacements.begin(), size_t{0});
+    rmm::device_uvector<vertex_t> allgathered_unique_cluster_ids(
+      displacements.back() + recvcounts.back(), handle.get_stream());
+    cugraph::device_allgatherv(handle.get_comms(),
+                               unique_cluster_ids.begin(),
+                               allgathered_unique_cluster_ids.begin(),
+                               recvcounts,
+                               displacements,
+                               handle.get_stream());
+
+    thrust::sort(handle.get_thrust_policy(),
+                 allgathered_unique_cluster_ids.begin(),
+                 allgathered_unique_cluster_ids.end());
+
+    allgathered_unique_cluster_ids.resize(
+      thrust::distance(allgathered_unique_cluster_ids.begin(),
+                       thrust::unique(handle.get_thrust_policy(),
+                                      allgathered_unique_cluster_ids.begin(),
+                                      allgathered_unique_cluster_ids.end())),
+      handle.get_stream());
+
+    detail::relabel_cluster_ids<vertex_t, multi_gpu>(
+      handle, allgathered_unique_cluster_ids, clustering, local_num_verts);
+
+  } else {
+    detail::relabel_cluster_ids<vertex_t, multi_gpu>(
+      handle, unique_cluster_ids, clustering, local_num_verts);
+  }
+
```

Comment:
LGTM, but please test it rigorously. 
Currently we don't have a test to check if the produced cluster ids are consecutive , starting from 0.

Reply:
You need to comment out the lines
https://github.com/rapidsai/cugraph/blob/2b1ac70272c2f9a08d688496c4ae88104acb2c62/cpp/src/community/leiden_impl.cuh#L606C1-L627C4
as it's used differently now.